### PR TITLE
[WIP][CR] Brain Cylinders as items you can talk to

### PR DESCRIPTION
**Pull request by Light-Wave**
Adds TALK_TO function to items, which lets you talk to an NPC that lives inside the item.
Adds brain cylinders and an NPC that lives inside them.

#### Summary

SUMMARY: Features "Adds Brain Cylinders, items that you can talk to. Also adds functionality for talking to items."

#### Purpose of change

- Adds functionality to talk to items (through C++)
- Adds item to talk to

#### Describe the solution

- Adds TALK_TO as an available use_action for items
  - When Talk_To is invoked, the code will check if the item contains data about an existing NPC
    - If not, create an NPC of id "<item name>_talker"
    - If it does, load the NPC from memory
  - Make sure the NPC has no hunger, thirst, or fatigue, and isn't sleeping
  - Open Dialogue with the NPC
  - Save changes to the NPC
    - // TODO: Make sure the NPC's inventory can not contain the item that contain the NPC
- Adds a Brain Cylinder item that can be talked to
- Adds NPC data for the NPC inside the brain cylinder
- Adds dialogue for the NPC
  - // WIP: Intro quest inspired by prototype cyborgs
  - // TODO: Make brain offer to teach skills and/or recepies if befriended
- // TODO: Possibly some JSON data about starting equipment, class, name list, material, deconstruction, background, etc
- // TODO: Documentation for TALK_TO
- // TODO: item image
- // TODO: Make item apear in world, possibly in mi-go related locations, and perhaps as dropp from mi-go surgeons
- // TODO: Merge item with living_brain_jar from Aftershock? Make Aftershock recipes accept this item?
- Make furniture able to talk
- Make Hub 01 intercom use this system instead.
  - // TODO: Update Hub1 mapgen to use new system

// TODO: Move JSON from mod folder to main JSON files
// TODO: Verify style

#### Describe alternatives you've considered

- Spawn an invisible NPC that imediately starts a dialogue and despawns directly after being talked to. Might be problematic if there is no space for them to spawn.
- Use a global NPC stored in the map or something instead of storing them in the item, would maybe prevent risk of infinite regress if you manage to get the item inside the NPCs inventory.
- Use the computer interface to create dialogue instead. Unfortunately, the computer interface seems to be hard-coded into C++ and not modifiable by JSON.

#### Testing

Spawning in the item and talking to it seems to work. Data is saved between talks. It is possible to trade with the NPC, which is unintended, but will be left in in case someone want to use that functionality for something like a Minecraft Equivalent Exhange [Transmutation Tablet](https://technicpack.fandom.com/wiki/Transmutation_Tablet).
Trading the Item to the NPC did not crash the game.

#### Additional context

brain1
brain2
brain3

### CODE REVIEW

**Comment by anothersimulacrum** [comment]
It would be preferred that you add talker_item (like talker)character/npc/avatar) instead of using an NPC.

**Comment by Leland** [comment]
Nice, feels very Futurama. Where in the CDDA world can you find one of these brain jars? The lab seems like a natural choice.

This also feels like something that would do great in Crazy Cataclysm!

**Comment by Amoebka** [comment]
This is most certainly a mi-go thing, not a futurama thing.

**Comment by Light-Wave** [comment]
It would indeed be a mi-go thing. I'm thinking having them spawn in mi-go scout towers and outposts, as well as possible drop from mi-go surgeons.

**Comment by ScampsAdams** [comment]
Furniture next? (intercom)

**Comment by Zireael07** [comment]
Is this a mod thing, or a core thing? When I read the title, I thought "that's cyberpunk [Aftershock] as hell" <3

File: data/mods/Brain_Cylinders/items.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[
+  {
+    "id": "brain_cylinder_1",
+    "type": "GENERIC",
+    "name": "Brain Cylinder",
+    "description": "This is a glass cylinder containing a brain. Strange organic-looking growths of resin caps both ends. There apears to be a button on the bottom.",
```
Comment (matskuman5) [inline data/mods/Brain_Cylinders/items.json:6]:
```suggestion
    "description": "This is a glass cylinder containing a brain. Strange organic-looking growths of resin cap both ends. There appears to be a button on the bottom.",
```

**Comment by Light-Wave** [comment]
@ScampsAdams That sounds like a great idea. That would allow people to easier make intercom-like furniture without having to hide a dummy NPC nearby.

@Zireael07 It is intended to be a core thing. This is mi-go tech. That being said, I think it should be possible to use these to create Frankenstein monsters from Aftershock

**Comment by I-am-Erk** [comment]
How hard would it be to make furniture able to talk as well as part of this?

**Comment by Light-Wave** [comment]
@I-am-Erk I believe it would be fairly easy to make furniture able to talk as well. I'll add it to my to-do list.

**Comment by aoae23** [comment]
Really exciting feature. As mentioned above, if you're up for it, you should apply it to the Hub 01 intercom NPC

**Comment by Light-Wave** [comment]
@anothersimulacrum I fear I do not have a deep enough understanding of the differences between different talker_ classes to know what differences there would be between a talker_item and a talker_npc. Would you care to elaborate what differences you have in mind?

@aoae23 I will attempt to apply it to the Hub 01 intercom, yes.

**Review by mlangsdorf — CHANGES_REQUESTED** [review]
NAK.  Please implement `talker_item` for this functionality

**Comment by mlangsdorf** [comment]
The `talker` class is a virtual abstract class that implements all the function calls used by existing dialogue conditions and dialogue effects.  Any child class of `talker` is guaranteed to work with all existing TALK_TOPICs.

The `talker` child classes all take an entity (currently player, avatar, or npc but it can be extended to anything) as an argument to their constructor, and create a `talker_` child class that can be passed to `npctalk.cpp` for dialogue handling.  This gets around a lot of the limitations of the current intercom workaround (there aren't any NPCs that need to be hidden, for instance) and expands the game to allow the player to talk to a brain cylinder, an intercom, or whatever else is needed.

Please read the talker classes (I think they're reasonably well documented) and ping me at mark.langsdorf@gmail.com or mlangsdorf on the dev Discord as you have more questions.

**Review by mlangsdorf — CHANGES_REQUESTED** [review]
I'm still seeing a lack of a `talker_item` class that is necessary for this to get merged.

**Comment by Light-Wave** [comment]
> I'm still seeing a lack of a `talker_item` class that is necessary for this to get merged.

**Comment by Light-Wave** [comment]
> I'm still seeing a lack of a `talker_item` class that is necessary for this to get merged.

I haven't gotten around to it, but intend to implement it.

**Pull request closed without merging**